File: app/routers/runs.py
"""Run lifecycle endpoints plus build/run lookup and logs."""
from __future__ import annotations

import asyncio
import json

from fastapi import APIRouter, HTTPException, Request
from fastapi.responses import PlainTextResponse, StreamingResponse

from ..auth import resolve_user
from ..responses import data, get_service, not_found, read_body
from ..serializers import build_dict, run_dict


def _bounded_int(params, key: str, default: int, minimum: int, message: str) -> int:
    """Parse an integer query param with a lower bound, or raise a 400 (never a bare 500).

    Mirrors the malformed-body handling: a non-integer or out-of-range value is
    caller error, so it maps to HTTP 400 in the Apify error shape rather than an
    uncaught ``ValueError`` that FastAPI would surface as a 500. ``minimum=1``
    yields "must be positive" semantics; ``minimum=0`` yields "must not be
    negative" semantics.
    """
    raw = params.get(key)
    if raw is None or raw == "":
        return default
    try:
        value = int(raw)
    except (TypeError, ValueError):
        raise HTTPException(status_code=400, detail=f"Query parameter '{key}' must be an integer.")
    if value < minimum:
        raise HTTPException(status_code=400, detail=message)
    return value

# Start-run lives under the actor prefixes (/v2/acts + /v2/actors).
actor_runs_router = APIRouter()

# Flat resources.
flat_router = APIRouter()


@actor_runs_router.post("/{actor_id}/runs")
async def start_run(actor_id: str, request: Request) -> object:
    svc = get_service(request)
    user = await resolve_user(request)
    actor = await svc.get_actor(actor_id, username=user)
    if actor is None:
        return not_found(f"Actor '{actor_id}' was not found.")
    raw = await read_body(request)
    try:
        run_input = json.loads(raw) if raw else {}
    except json.JSONDecodeError:
        run_input = {}
    params = request.query_params
    # Validate every query param BEFORE starting the run so a bad value returns a
    # 400 without spawning a run. memory/timeout must be positive integers (a
    # zero/negative memory would otherwise silently disable the container memory
    # cap); waitForFinish may be 0.
    options = {
        "build": params.get("build", "latest"),
        "memoryMbytes": _bounded_int(
            params, "memory", 1024, minimum=1, message="Query parameter 'memory' must be positive."
        ),
        "timeoutSecs": _bounded_int(
            params, "timeout", 300, minimum=1, message="Query parameter 'timeout' must be positive."
        ),
    }
    wait_secs = min(
        _bounded_int(
            params, "waitForFinish", 0, minimum=0, message="Query parameter 'waitForFinish' must not be negative."
        ),
        60,
    )

    run = await svc.start_run(actor_id, run_input, options)

    deadline = wait_secs
    while deadline > 0:
        current = await svc.get_run(run.id)
        if current.status in {"SUCCEEDED", "FAILED", "ABORTED", "TIMED-OUT"}:
            run = current
            break
        await asyncio.sleep(0.5)
        deadline -= 0.5
    else:
        run = await svc.get_run(run.id)
    return data(run_dict(run), status_code=201)


@actor_runs_router.get("/{actor_id}/runs")
async def list_runs(actor_id: str, request: Request) -> object:
    svc = get_service(request)
    user = await resolve_user(request)
    runs = await svc.list_runs(actor_id, username=user)
    items = [run_dict(r) for r in runs]
    return data({"total": len(items), "count": len(items), "items": items})


@flat_router.get("/v2/actor-runs/{run_id}")
async def get_run(run_id: str, request: Request) -> object:
    svc = get_service(request)
    user = await resolve_user(request)
    run = await svc.get_run(run_id, username=user)
    if run is None:
        return not_found(f"Run '{run_id}' was not found.")
    return data(run_dict(run))


@flat_router.post("/v2/actor-runs/{run_id}/abort")
async def abort_run(run_id: str, request: Request) -> object:
    svc = get_service(request)
    user = await resolve_user(request)
    run = await svc.abort_run(run_id, username=user)
    if run is None:
        return not_found(f"Run '{run_id}' was not found.")
    return data(run_dict(run))


@flat_router.post("/v2/actor-builds/{build_id}/abort")
async def abort_build(build_id: str, request: Request) -> object:
    svc = get_service(request)
    user = await resolve_user(request)
    build = await svc.abort_build(build_id, username=user)
    if build is None:
        return not_found(f"Build '{build_id}' was not found.")
    return data(build_dict(build))


@flat_router.get("/v2/actor-builds/{build_id}")
async def get_build(build_id: str, request: Request) -> object:
    svc = get_service(request)
    user = await resolve_user(request)
    build = await svc.get_build(build_id, username=user)
    if build is None:
        return not_found(f"Build '{build_id}' was not found.")
    return data(build_dict(build))


# Logs are dynamic and must never be cached. no-store also opts these
# responses out of the browser's same-URL cache lock, which would otherwise
# queue a re-opened log view behind a still-open (never-ending, for a warm
# standby run) earlier stream to the same URL and render it empty forever.
_LOG_NO_STORE = {"Cache-Control": "no-store"}


@flat_router.get("/v2/logs/{job_id}")
async def get_log(job_id: str, request: Request) -> PlainTextResponse:
    svc = get_service(request)
    user = await resolve_user(request)
    build = await svc.get_build(job_id, username=user)
    if build is not None:
        return PlainTextResponse(build.log or "", headers=_LOG_NO_STORE)
    run = await svc.get_run(job_id, username=user)
    if run is not None:
        # A warm standby run's log lives only in its container until teardown
        # persists it; fetch it live so the log is not empty while RUNNING.
        live = await svc.standby.live_container_log(run)
        if live is not None:
            return PlainTextResponse(live, headers=_LOG_NO_STORE)
        return PlainTextResponse(run.log or "", headers=_LOG_NO_STORE)
    return PlainTextResponse("", status_code=404)


_TERMINAL_STATUSES = {"SUCCEEDED", "FAILED", "ABORTED", "TIMED-OUT"}
_STREAM_POLL_SECS = 0.25


@flat_router.get("/v2/logs/{job_id}/stream")
async def stream_log(job_id: str, request: Request):
    svc = get_service(request)
    user = await resolve_user(request)

    # Resolve the job's kind once up front (a build and a run never share an id):
    # this doubles as the unknown / cross-user 404 guard exactly like the one-shot
    # endpoint, and lets each poll tick below re-fetch only the relevant object.
    is_build = await svc.get_build(job_id, username=user) is not None
    if not is_build and await svc.get_run(job_id, username=user) is None:
        return PlainTextResponse("", status_code=404)

    async def _status_and_log() -> tuple[bool, str]:
        """Return ``(is_terminal, stored_log)`` for the caller's build/run."""
        job = (
            await svc.get_build(job_id, username=user)
            if is_build
            else await svc.get_run(job_id, username=user)
        )
        if job is None:
            return True, ""
        return job.status in _TERMINAL_STATUSES, (job.log or "")

    async def _live_standby_log() -> str | None:
        """Live container log for a warm standby run, else None (see get_log)."""
        if is_build:
            return None
        run = await svc.get_run(job_id, username=user)
        if run is None:
            return None
        return await svc.standby.live_container_log(run)

    async def _tail():
        offset = 0
        while True:
            buf = svc.read_log_buffer(job_id)
            if buf is None:
                # Standby runs never create a live buffer; their in-container
                # log grows monotonically and teardown persists that same text
                # (plus a trailing note), so the offset stays consistent when
                # the terminal drain below switches to the stored log.
                buf = await _live_standby_log()
            if buf is not None and len(buf) > offset:
                yield buf[offset:]
                offset = len(buf)
            terminal, stored = await _status_and_log()
            if terminal:
                # Final drain: emit anything appended to the buffer since the last
                # read, then any stored-log tail (e.g. a post-run import error that
                # the live stream never carried), so a client tailing right at the
                # finish still receives the end of the log before the stream closes.
                buf = svc.read_log_buffer(job_id)
                if buf is not None and len(buf) > offset:
                    yield buf[offset:]
                    offset = len(buf)
                if len(stored) > offset:
                    yield stored[offset:]
                return
            await asyncio.sleep(_STREAM_POLL_SECS)

    return StreamingResponse(_tail(), media_type="text/plain", headers=_LOG_NO_STORE)

File: src/routes/runs.js
/** Run lifecycle endpoints plus build/run lookup and logs. */
import { resolveUser } from '../auth.js';
import { TERMINAL_STATUSES } from '../constants.js';
import { data, notFound, readBody, response, textResponse } from '../http.js';
import { boundedInt } from '../pagination.js';
import { buildDict, runDict } from '../serializers.js';

// Logs are dynamic and must never be cached. no-store also opts these
// responses out of the browser's same-URL cache lock, which would otherwise
// queue a re-opened log view behind a still-open (never-ending, for a warm
// standby run) earlier stream to the same URL and render it empty forever.
const LOG_NO_STORE = [['cache-control', 'no-store']];

const STREAM_POLL_MS = 250;

function sleep(ms) {
    return new Promise((resolve) => {
        setTimeout(resolve, ms);
    });
}

/** Start-run lives under the actor prefixes (/v2/acts + /v2/actors). */
export function registerActorRunRoutes(router, prefix) {
    router.add('POST', `${prefix}/:actorId/runs`, async (ctx, { actorId }) => {
        const svc = ctx.service;
        const user = await resolveUser(ctx);
        const actor = svc.getActor(actorId, user);
        if (!actor) return notFound(`Actor '${actorId}' was not found.`);
        const raw = await readBody(ctx);
        let runInput = {};
        if (raw.length) {
            try {
                runInput = JSON.parse(raw.toString('utf8'));
            } catch {
                runInput = {};
            }
        }
        // Validate every query param BEFORE starting the run so a bad value
        // returns a 400 without spawning a run. memory/timeout must be
        // positive integers (a zero/negative memory would otherwise silently
        // disable the container memory cap); waitForFinish may be 0.
        const options = {
            build: ctx.query.get('build') || 'latest',
            memoryMbytes: boundedInt(ctx.query, 'memory', 1024, 1, "Query parameter 'memory' must be positive."),
            timeoutSecs: boundedInt(ctx.query, 'timeout', 300, 1, "Query parameter 'timeout' must be positive."),
        };
        const waitSecs = Math.min(
            boundedInt(ctx.query, 'waitForFinish', 0, 0, "Query parameter 'waitForFinish' must not be negative."),
            60,
        );

        let run = await svc.startRun(actorId, runInput, options);

        let deadline = waitSecs;
        while (deadline > 0) {
            const current = svc.getRun(run.id);
            if (TERMINAL_STATUSES.has(current.status)) {
                run = current;
                break;
            }
            await sleep(500);
            deadline -= 0.5;
        }
        run = svc.getRun(run.id);
        return data(runDict(run), 201);
    });

    router.add('GET', `${prefix}/:actorId/runs`, async (ctx, { actorId }) => {
        const svc = ctx.service;
        const user = await resolveUser(ctx);
        const items = svc.listRuns(actorId, user).map(runDict);
        return data({ total: items.length, count: items.length, items });
    });
}

export function registerFlatRunRoutes(router) {
    router.add('GET', '/v2/actor-runs/:runId', async (ctx, { runId }) => {
        const user = await resolveUser(ctx);
        const run = ctx.service.getRun(runId, user);
        if (!run) return notFound(`Run '${runId}' was not found.`);
        return data(runDict(run));
    });

    router.add('POST', '/v2/actor-runs/:runId/abort', async (ctx, { runId }) => {
        const user = await resolveUser(ctx);
        const run = await ctx.service.abortRun(runId, user);
        if (!run) return notFound(`Run '${runId}' was not found.`);
        return data(runDict(run));
    });

    router.add('POST', '/v2/actor-builds/:buildId/abort', async (ctx, { buildId }) => {
        const user = await resolveUser(ctx);
        const build = ctx.service.abortBuild(buildId, user);
        if (!build) return notFound(`Build '${buildId}' was not found.`);
        return data(buildDict(build));
    });

    router.add('GET', '/v2/actor-builds/:buildId', async (ctx, { buildId }) => {
        const user = await resolveUser(ctx);
        const build = ctx.service.getBuild(buildId, user);
        if (!build) return notFound(`Build '${buildId}' was not found.`);
        return data(buildDict(build));
    });

    router.add('GET', '/v2/logs/:jobId', async (ctx, { jobId }) => {
        const svc = ctx.service;
        const user = await resolveUser(ctx);
        const build = svc.getBuild(jobId, user);
        if (build) return textResponse(build.log ?? '', 200, LOG_NO_STORE);
        const run = svc.getRun(jobId, user);
        if (run) {
            // A warm standby run's log lives only in its container until
            // teardown persists it; fetch it live so the log is not empty
            // while RUNNING.
            const live = await svc.standby.liveContainerLog(run);
            if (live !== null) return textResponse(live, 200, LOG_NO_STORE);
            return textResponse(run.log ?? '', 200, LOG_NO_STORE);
        }
        return textResponse('', 404);
    });

    router.add('GET', '/v2/logs/:jobId/stream', async (ctx, { jobId }) => {
        const svc = ctx.service;
        const user = await resolveUser(ctx);

        // Resolve the job's kind once up front (a build and a run never share
        // an id): this doubles as the unknown / cross-user 404 guard exactly
        // like the one-shot endpoint, and lets each poll tick below re-fetch
        // only the relevant object.
        const isBuild = svc.getBuild(jobId, user) !== null;
        if (!isBuild && svc.getRun(jobId, user) === null) {
            return textResponse('', 404);
        }

        const statusAndLog = () => {
            const job = isBuild ? svc.getBuild(jobId, user) : svc.getRun(jobId, user);
            if (!job) return { terminal: true, stored: '' };
            return { terminal: TERMINAL_STATUSES.has(job.status), stored: job.log ?? '' };
        };

        const liveStandbyLog = async () => {
            if (isBuild) return null;
            const run = svc.getRun(jobId, user);
            if (!run) return null;
            return svc.standby.liveContainerLog(run);
        };

        async function* tail() {
            let offset = 0;
            for (;;) {
                let buffer = svc.readLogBuffer(jobId);
                if (buffer === null) {
                    // Standby runs never create a live buffer; their
                    // in-container log grows monotonically and teardown
                    // persists that same text (plus a trailing note), so the
                    // offset stays consistent when the terminal drain below
                    // switches to the stored log.
                    buffer = await liveStandbyLog();
                }
                if (buffer !== null && buffer.length > offset) {
                    yield buffer.slice(offset);
                    offset = buffer.length;
                }
                const { terminal, stored } = statusAndLog();
                if (terminal) {
                    // Final drain: emit anything appended to the buffer since
                    // the last read, then any stored-log tail (e.g. a
                    // post-run import error the live stream never carried),
                    // so a client tailing right at the finish still receives
                    // the end of the log before the stream closes.
                    const finalBuffer = svc.readLogBuffer(jobId);
                    if (finalBuffer !== null && finalBuffer.length > offset) {
                        yield finalBuffer.slice(offset);
                        offset = finalBuffer.length;
                    }
                    if (stored.length > offset) {
                        yield stored.slice(offset);
                    }
                    return;
                }
                await sleep(STREAM_POLL_MS);
            }
        }

        return response({
            status: 200,
            headers: [['content-type', 'text/plain; charset=utf-8'], ...LOG_NO_STORE],
            stream: tail(),
        });
    });
}
